include/zephyr/drivers/flash.h

```diff
@@ -422,6 +427,67 @@ static inline const struct flash_parameters *z_impl_flash_get_parameters(const s
 	return api->get_parameters(dev);
 }
 
+/**
+ *  @brief Execute flash extended operation on given device
+ *
+ *  Besides of standard flash operations like write or erase, flash controllers
+ *  also support additional features like write protection or readout
+ *  protection. These features are not available in every flash controller,
+ *  what's more controllers can implement it in a different way.
+ *
+ *  It doesn't make sense to add a separate flash API function for every flash
+ *  controller feature, because it could be unique (supported on small number of
+ *  flash controllers) or the API won't be able to represent the same feature on
+ *  every flash controller.
+ *
+ *  @param dev Flash device
+ *  @param code Operation which will be executed on the device.
+ *  @param in Pointer to input data used by operation. If operation doesn't
+ *            need any input data it could be NULL.
+ *  @param out Pointer to operation output data. If operation doesn't produce
+ *             any output it could be NULL.
+ *
+ *  @retval 0 on success.
+ *  @retval -ENOTSUP if given device doesn't support extended operation.
+ *  @retval -ENOSYS if support for extended operations is not enabled in Kconfig
+ *  @retval negative value on extended operation errors.
+ */
+__syscall int flash_ex_op(const struct device *dev, uint16_t code,
+			  const uintptr_t in, void *out);
```

Comment:
One more question: why `out` is `void *` not `uintptr_t` ?

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/53441#discussion_r1097282613

> I was thinking about direct passing integer parameters via registers, for example when you have some function like:
> 
> ```
> flash_ex_op(dev, SET_SLEEP_LEVEL, x, NULL) // x is level in range 0-N
> flash_ex_op(dev, GET_POWER_INFO_BLOCK, x, p) // where x is info index that we want, and p is x specific type pointer to structure where the info will be placed
> ```
> 
> to avoid needing to do `copyin` of whatever `in` points, to kernel space, before accessing it, when it is not needed - for example to process single int param. `out` should be `void *`, `intptr_t` does not make sense here);

You mentioned that `out` should be `void *`.

Reply:
OK. I am dumb. This is OK, but next time when I write something like
```
intptr_t does not make sense here
```
can somebody ask why?

Reply:
It's clear for me why `uintptr_t` is not better than `void *` for `out` parameter, but I can ask :P

Reply:
Yeah. I work to many things, forgot my previous argument and I just got jumped by another "clever" Idea for `out`; now I remember I had it before and why I have dropped it.